src/core/server/legacy/plugins/get_nav_links.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { LegacyUiExports, LegacyNavLink, LegacyPluginSpec } from '../types';
+
+const REMOVE_FROM_ARRAY: LegacyNavLink[] = [];
+
+function getUiAppsNavLinks({ uiAppSpecs = [] }: LegacyUiExports, pluginSpecs: LegacyPluginSpec[]) {
+  return uiAppSpecs.flatMap(spec => {
+    if (!spec) {
+      return REMOVE_FROM_ARRAY;
+    }
+
+    const id = spec.id;
+
+    if (!id) {
+      throw new Error('Every app must specify an id');
+    }
+
+    if (spec.pluginId && !pluginSpecs.some(plugin => plugin.getId() === spec.pluginId)) {
+      throw new Error(`Unknown plugin id "${spec.pluginId}"`);
+    }
+
+    const listed = typeof spec.listed === 'boolean' ? spec.listed : true;
+
+    if (spec.hidden || !listed) {
+      return REMOVE_FROM_ARRAY;
+    }
+
+    return {
+      id,
+      category: spec.category,
+      title: spec.title,
+      order: typeof spec.order === 'number' ? spec.order : 0,
+      icon: spec.icon,
+      euiIconType: spec.euiIconType,
+      url: spec.url || `/app/${id}`,
+      linkToLastSubUrl: spec.linkToLastSubUrl,
```

Comment:
seems to be optional and should repeat this logic https://github.com/elastic/kibana/pull/57542/files#diff-d1ea0ed9e7f4c8bacd7b2ecc4453fa05R71
I'd suggest normalizing spec content, even though this code will be removed soon-ish

Reply:
> I'd suggest normalizing spec content

These are two distinct types though, `LegacyNavLinkSpec` and `LegacyAppSpec`, with different fields and optionals on field (because, 'reasons' I guess). Not really confident on touching these things?

Reply:
Can try to extract a mapping function that would accept either of those though